### PR TITLE
HTTPHandler abstraction fixing

### PR DESCRIPTION
**Pull request by sawradip**
### **User description**
Currently, in `sdk.py` HTTP (get/post/...) calls are handled from different functions. There should be a centralised class `HttpHandler` to handle all http calls, all interaction to backend api. So I am introducing a class proposing a change(partial for now).

```

class HttpHandler():
    def __init__(self, base_url:str, api_key:str):
        self.http_client = HttpClient(base_url)
        self.http_client.headers.update(
            {"Content-Type": "application/json", "x-api-key": api_key}
        )

    def get_action_schemas(self, app_unique_ids:list[str]= [],  use_case: str = "", limit: int = 0):

        params = {}
        if use_case:
            params["use_case"] = use_case
            if limit:
                params["limit"] = limit
            if len(app_unique_ids) != 1:
                raise ValueError("Use case should be provided with exactly one app.")

        # Cannot apply limit without usecase
        elif limit != 0:
            raise ValueError("Limit can be only mentioned with Use case")
        
        if app_unique_ids:
            params["appNames"]= ",".join(app_unique_ids )

        resp = self.http_client.get(f"v1/actions", params=params)

        return resp.json()
```

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced `HttpHandler` class in `http_client.py` to centralize HTTP calls.
- Integrated `HttpHandler` into `ComposioCore` class in `sdk.py`.
- Refactored `get_list_of_actions` method in `sdk.py` to use `HttpHandler`.
- Added a Jupyter notebook `refactor_debug.ipynb` for debugging and testing the refactored code.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>http_client.py</strong><dd><code>Introduce `HttpHandler` class for centralized HTTP calls.</code></dd></summary>
<hr>
      
composio/sdk/http_client.py

<li>Introduced <code>HttpHandler</code> class to centralize HTTP calls.<br> <li> Added method <code>get_action_schemas</code> to <code>HttpHandler</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/50/files#diff-75b82c6f46d40455ae42025d65d4ddc1c65bdcac54423c0a845780392ffc8f30">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>sdk.py</strong><dd><code>Integrate `HttpHandler` into `ComposioCore` and refactor methods.</code></dd></summary>
<hr>
      
composio/sdk/sdk.py

<li>Integrated <code>HttpHandler</code> into <code>ComposioCore</code>.<br> <li> Refactored <code>get_list_of_actions</code> to use <code>HttpHandler</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/50/files#diff-f48086c75a8ebff7ac889308e11327f687917a7985c7a84b1a17ed08715c5a33">+134/-1</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>refactor_debug.ipynb</strong><dd><code>Add notebook for debugging and testing refactored code.</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
refactor_debug.ipynb

<li>Added notebook for debugging and testing refactored code.<br> <li> Demonstrated usage of <code>HttpHandler</code> in the notebook.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/50/files#diff-0ec9f006b62e2a5c7cd6d24f888d7f2b9d2779b4d3ad038a5e1551a3b1b9cfd1">+1657/-0</a></td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: refactor_debug.ipynb [hunk 1]
```diff
@@ -0,0 +1,1657 @@
+{
```
Comment (kaavee315) [inline refactor_debug.ipynb:1]: Rm this file?

File: composio/sdk/sdk.py [hunk 2]
```diff
@@ -312,84 +313,60 @@ def get_app(self, app_name: str):
...
-        use_case: str = None,
-        actions: list[Action] = None,
-        tags: list[Union[str, Tag]] = None,
-        limit: int = None,
+        apps: list[App] = [],
+        actions: list[Action] = [],
+        use_case: str = "",
+        tags: list[Union[str, Tag]] = [],
+        limit: int = 0,
     ) -> list:
```
Comment (kaavee315) [inline composio/sdk/sdk.py:321]: Are we missing empty apps case here -> get all apps and their actions?

File: composio/sdk/sdk.py [hunk 3]
```diff
@@ -312,84 +313,60 @@ def get_app(self, app_name: str):
...
+
+            elif action_schema["name"] in action_unique_ids:
+                filtered_actions.append(action_schema)
+                important_actions.append(action_schema)
+
+        if len(important_actions) > 5 and Tag.ALL not in tags:
+            return important_actions
         else:
-            raise ValueError("Either apps or actions must be provided")
+            return filtered_actions
```
Comment (kaavee315) [inline composio/sdk/sdk.py:369]: What if a proper filter is given like Tag.Repo in that case also above condition will return important _actions. Fix it.

**Review by kaavee315 — CHANGES_REQUESTED** [review]

File: refactor_debug.ipynb (hunk 1 above)
Comment (sawradip) [reply refactor_debug.ipynb:1]: Yeah..sure. This was for the reviewer's testing...as I couldn't add tests yet. @kaavee315

File: composio/sdk/sdk.py (hunk 3 above)
Comment (sawradip) [reply composio/sdk/sdk.py]: Solved.

File: composio/sdk/sdk.py (hunk 2 above)
Comment (sawradip) [reply composio/sdk/sdk.py:321]: Solved.

File: composio/sdk/sdk.py [hunk 4]
```diff
@@ -312,84 +313,75 @@ def get_app(self, app_name: str):
...
+            elif action_schema["name"] in action_unique_ids:
+                filtered_actions.append(action_schema)
+                important_actions.append(action_schema)
+
+        if (tags == [Tag.IMPORTANT]) or (
+            len(important_actions) > 5
+            and tag_values == [Tag.IMPORTANT.value[1]]
+            and all_relevent_app_ids != []
+        ):
+            return important_actions
```
Comment (kaavee315) [inline composio/sdk/sdk.py:382]: Basically if I give tag = Tag.REPOS, then this wouldn't be invoked right?

**Comment by kaavee315** [comment]
Let's add the following cases:
1. One app, multiple apps with/without tags
2. One action, multiple actions.

**Review by kaavee315 — APPROVED** [review]
LGTM Mdoulo testing:
1. One app, multiple apps with/without tags
2. One action, multiple actions.
3. App with use case with/out limit

File: composio/sdk/sdk.py (hunk 4 above)
Comment (sawradip) [reply composio/sdk/sdk.py:382]: No, This is only for the case when Tag.Important is specifically mentioned, or no tag is mentioned with atleast one app. @kaavee315

**Comment by angrybayblade** [comment]
Closing in favour of - https://github.com/SamparkAI/composio/pull/53

**Pull request closed without merging**